### ISSUE TITLE
Latest dev release actually slower than 0.5

### ISSUE THREAD

**Issue by nryant** [tensorflow/tensorflow#548]
After updating TensorFlow to the most recent source yesterday (I'm at b1cabed4e60015602dacd66ea39d419db50c3e1b), I've noticed that while GPU utilization frequently appears much higher in nividia-smi than in prior releases, my actual code is much slower. Some sequence to sequence models I was training began taking 3-4 times as long per step, despite GPU utilization hovering between 60 and 99%, which is much higher than I have observed in the past. As I have code for benchmarking fully connected feedforward networks on MNIST in various frameworks, I dusted that off and, again, slower. Previously, training a network with three hidden layers of 2,048 rectified linear units + dropout (input + hidden) took 1.78 seconds per epoch (averaged over 10 epochs)  when trained using vanilla SGD with momentum and a minibatch size of 256. That is now up to 65.2 seconds. This holds across different combinations of hidden layer sizes and minibatch sizes. On the other hand, convolutional net performance does not seem to be affected as when I run Soumith's convolutional net benchmarks, I get numbers close to what he originally reported using the same test setup.

So to summarize, I've been recompiling TensorFlow regularly (every few days since its release) and after the most recent compile noticed a quite substantial performance hit for vanilla fully connected feedforward and recurrent architectures, but not for convolutional networks. This is all with TensorFlow running on a Titan X with no other processes running and using the most recent versions of CUDA, cuDNN (well, I have v3 installed, not the release candidate for v4), cuBLAS, etc.

**Comment by zheng-xq** [comment]
@nryant, we will look into this problem. But to make sure we are looking at
the same problem, could you provide some more reproducing details? It will
be very helpful if you can point us to a model either comes with TF, or
written by you, that is confirmed to be slower on TF-0.6 and faster on
TF-0.5.

**Comment by nryant** [comment]
Over the weekend I can put up a repo on github containing code to reproduce the fully connected model I mentioned. I'll also try downgrading to a previous version to see if I can replicate the timings I recorded a few weeks ago. Would that be sufficient?

**Comment by zheng-xq** [comment]
That would be very helpful! Thanks.

**Comment by benoitsteiner** [comment]
We used to compile with our own version of Eigen. We started using the upstream version of Eigen last week, and it turns out that a couple optimizations are missing. I'll fix the upstream version as soon as possible to get back to where we were performance-wise.

**Comment by ebrevdo** [comment]
Benoit, is the correct tanh implementation upstream?

**Comment by nryant** [comment]
The promised repo is at 

https://github.com/nryant/tensorflow_mnist_examples/tree/master

and the benchmark located under the feedforward_bench directory therein. When run, feedforward_bench/bench.py will spit out to a file average time to complete an epoch (averaged over ten epochs) over MNIST for a fully connected feedforward model with 3 layers and varying layer sizes and minibatch sizes. I've included output from runs using the official 0.60 release and the more recent release cited upthread (see the .txt files). I've also included log files for each run containing device placement for the operations.

As you can see, the more recent release is FAR slower, though rather interestingly nvidia-smi shows much higher GPU utilization for it.

**Comment by JianbangZ** [comment]
is the fix upstream?

**Comment by benoitsteiner** [comment]
I have cheked in several improvements to the upstream eigen repository such as [this one] (https://bitbucket.org/eigen/eigen/commits/a0661a2bb16512dc4359c243db1dbeb9bc32b266). I have also updated the TensorFlow codebase to pull a recent version of Eigen that contains these improvements. This should fix the performance regression issue.

**Comment by nryant** [comment]
As of d1b8333effdcb031e6e34a2835a2f1c877fdd79b, this performance issue appears resolved. See:

https://github.com/nryant/tensorflow_mnist_examples/blob/master/feedforward_bench/d1b8333_timings.txt

**Comment by benoitsteiner** [comment]
Thanks for verifying the fixes. I'm closing this issue.
